### ISSUE TITLE
backend consul uses http when CONSUL_HTTP_SSL=true and scheme=https

### ISSUE THREAD

**Issue by transferkraM** [hashicorp/terraform#23688]
### Terraform Version

```
Terraform v0.12.18
+ provider.consul v2.6.1
```

### Terraform Configuration Files

```hcl
terraform {
  backend "consul" {
    address = "consul.dns.name:443"
    scheme  = "https"
    path    = "tfstate"
  }
}
```

### Debug Output

```
2019/12/16 16:03:15 [TRACE] Meta.Backend: instantiated backend of type *consul.Backend
2019/12/16 16:03:15 [DEBUG] checking for provider in "."
2019/12/16 16:03:15 [DEBUG] checking for provider in "/home/<user>/.tfenv/versions/0.12.18"
2019/12/16 16:03:15 [DEBUG] checking for provider in ".terraform/plugins/linux_amd64"
2019/12/16 16:03:15 [DEBUG] found provider "terraform-provider-consul_v2.6.1_x4"
2019/12/16 16:03:15 [DEBUG] found valid plugin: "consul", "2.6.1", "<...>"
2019/12/16 16:03:15 [DEBUG] checking for provisioner in "."
2019/12/16 16:03:15 [DEBUG] checking for provisioner in "/home/<user>/.tfenv/versions/0.12.18"
2019/12/16 16:03:15 [DEBUG] checking for provisioner in ".terraform/plugins/linux_amd64"
2019/12/16 16:03:15 [TRACE] Meta.Backend: backend *consul.Backend does not support operations, so wrapping it in a local backend
2019/12/16 16:03:15 [INFO] backend/local: starting Plan operation
2019/12/16 16:03:15 [TRACE] backend/local: requesting state manager for workspace "default"
2019/12/16 16:03:15 [TRACE] backend/local: requesting state lock for workspace "default"
2019/12/16 16:03:15 [INFO] created consul lock session <...>
Acquiring state lock. This may take a few moments...
2019/12/16 16:03:15 [TRACE] backend/local: reading remote state for workspace "default"
2019/12/16 16:03:15 [TRACE] backend/local: retrieving local state snapshot for workspace "default"
2019/12/16 16:03:15 [TRACE] backend/local: building context for current working directory
2019/12/16 16:03:15 [TRACE] terraform.NewContext: starting
2019/12/16 16:03:15 [TRACE] terraform.NewContext: resolving provider version selections
2019/12/16 16:03:16 [TRACE] terraform.NewContext: loading provider schemas
2019/12/16 16:03:16 [TRACE] LoadSchemas: retrieving schema for provider type "consul"
2019-12-16T16:03:16.004+0100 [INFO]  plugin: configuring client automatic mTLS
<...>
```

```
2019-12-16T16:03:16.449+0100 [DEBUG] plugin.terraform-provider-consul_v2.6.1_x4: 2019/12/16 16:03:16 [INFO] Consul Client configured with address: 'consul.dns.name:443', scheme: 'http', datacenter: '', insecure_https: 'false'
```

Environment variables exported:
* CONSUL_HTTP_SSL=true
* CONSUL_HTTP_ADDR=consul.dns.name:443
* CONSUL_HTTP_TOKEN=...

### Error Output

```
Error: Failed to read token '111ede88-c395-09a0-bc8e-d9f403b6e82a': Unexpected response code: 400 (Client sent an HTTP request to an HTTPS server.
)
```

### Expected Behavior

consul backend should use SSL if
* backend configuration `scheme` is set to `https`
* `CONSUL_HTTP_SSL=true` is exported

This functions properly if `CONSUL_HTTP_ADDR=https://consul.dns.name` is exported.

### Actual Behavior

consul backend uses `http://` scheme.

**Comment by fairlight12** [comment]
I've got the same error
🗿 terraform version
Terraform v0.12.24

**Comment by remilapeyre** [comment]
Hi @transferkraM and @fairlight12, thanks for reporting this issue.

I've try with both the latest version of Terraform and v0.12.24 but could not reproduce the issue:

```
$ cat example.tf                    
terraform {
  backend "consul" {
    address = "consul.lenstra.fr:8501"
    scheme  = "https"
    path    = "tfstate"
  }
}
$ export CONSUL_HTTP_SSL=true       
$ ~/Downloads/terraform_0.12.24 init

Initializing the backend...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
$ ~/Downloads/terraform_0.12.24 apply

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
$ env | grep CONSUL 
CONSUL_HTTP_SSL=true
```

It seems like the correct protocol has been used. Am I missing something?

**Comment by transferkraM** [comment]
Hi @remilapeyre.

The initial description might be insufficient.

I still have the issue, when the environment variable `CONSUL_HTTP_ADDR` is set without an explicit scheme. For me this is required to allow ansible-playbook to use the consul_kv lookup plugin (The same with vault lookup plugin.). Ansible strictly supports documented environment variables and fails if CONSUL_HTTP_ADDR has a scheme attached. TL;DR:

```bash
terraform version
```

```
Terraform v0.13.0
```

* Functions properly:

```bash
env|grep -v TOKEN|grep CONSUL
CONSUL_HTTP_SSL=true
```

* Still fails (required (for me) with Ansible)

```bash
env|grep -v TOKEN|grep CONSUL
CONSUL_HTTP_SSL=true
CONSUL_HTTP_ADDR=consul.dns.name:443
```

**Comment by remilapeyre** [comment]
This is weird, it seems to work on my end:

```
$ cat example.tf                                                                
terraform {
  backend "consul" {
    path = "tfstate"
  }

  required_providers {
    null = {
      source  = "hashicorp/null"
      version = "~> 2.1.2"
    }
  }
}

resource "null_resource" "test2" {
  provisioner "local-exec" {
    command = "sleep 1"
  }
}
$ env | grep CONSUL                                                             
$ terraform version                                                             
Terraform v0.13.0
$ CONSUL_HTTP_SSL=true CONSUL_HTTP_ADDR=consul.lenstra.fr:8501 terraform init

Initializing the backend...

Successfully configured the backend "consul"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Finding hashicorp/null versions matching "~> 2.1.2"...
- Installing hashicorp/null v2.1.2...
- Installed hashicorp/null v2.1.2 (signed by HashiCorp)

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
$ CONSUL_HTTP_SSL=true CONSUL_HTTP_ADDR=consul.lenstra.fr:8501 terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.test2 will be created
  + resource "null_resource" "test2" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.test2: Creating...
null_resource.test2: Provisioning with 'local-exec'...
null_resource.test2 (local-exec): Executing: ["/bin/sh" "-c" "sleep 1"]
null_resource.test2: Creation complete after 1s [id=6671518131037784894]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

**Comment by transferkraM** [comment]
I can confirm your setup. In my setup I still have the requirement to add 

```
provider "consul" {
  scheme = "https"
}
```

whenever I use consul related resources, and this brings me to: This issue is not related to terraform and init itself, but to the consul resources.

I'd close this here, because this issues description talks about consul **backend** rather than resources plugin.

**Comment by remilapeyre** [comment]
Oh, I see , thanks for looking into this.

The Consul backend and the Consul provider are currently separated so it is indeed possible that the behaviour differs between the two.

We had issues about the configuration of the provider in the past and while we fixed some (I think the last time was in the 2.6.1 releasse), it's possible that their is still some edge-cases that we haven't fixed.

Can you open a new issue at https://github.com/terraform-providers/terraform-provider-consul so we fix it there?

**Comment by remilapeyre** [comment]
This was an issue with the Consul Terraform provider and not the Consul backend in Terraform core. It has been fixed in the provider.

**Comment by remilapeyre** [comment]
Hi, this has been fixed in https://github.com/hashicorp/terraform-provider-consul and the fix was released in v2.10.0. I think we can close this issue now :)

**Comment by transferkraM** [comment]
Confirmed. Sorry, didn't find the time for testing/validating.
